XSD changes:

b-gyula opened the pull request:
- Do not allow other root element than `databaseChangeLog`
- added documentation for some elements
- Fix: addNotNullConstraint@schemaName + @catalogName was missing
- attributeGroups used only once merged into the elements they are used
- New attributeGroups for common attributes: schemaNameAttribute
- integerExp made based on xsd:long and non-negative
- property@name restricted according to the documentation site
- required string parameters made `nonEmptyString` to avoid SQL error for definitions like
-`rowCount@expectedRows`, `setTableRemarks@remarks` made mandatory

Code changes:
- simple clear error message if the root element is other than `databaseChangeLog`: databaseChangeLog expected as root element
- 'uniqueConstraint' was misspelled## Impact

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
<!--  Maintainers only: Mandatory Labels to add to a PR

At least one of the labels must be added to the PR before it's merged. If no label is provided the workflow will fail and you will not be able to merge the PR. After the label is added it re-runs the `Pull Request Labels / label (pull_request)` and gives a green check. 

`skipReleaseNotes`   - Don't show up on the Draft Release Notes page
`notableChanges`     - Any notable changes
`TypeEnhancement`    - New features
`TypeTest`           - New Test features
`TypeBug`            - bug fixes
`breakingChanges`    - any breaking changes
`APIBreakingChanges` - any API breaking changes
`sdou`               - Security, Driver and Other Updates -dependabot PR's
`newContributors`    - New Contributors 

-->


## Description

<!--
A clear and concise description of the change being made.  

- Introduce what was/will be done in the title
  - Titles show in release notes and search results, so make them useful
  - Use verbs at the beginning of the title, such as "fix", "implement", "improve", "update", and "add" 
  - Be specific about what was fixed or changed
  - Good Example: `Fix the --should-snapshot-data CLI parameter to be preserved when the --data-output-directory property is not specified in the command.`
  - Bad Example: `Fixed --should-snapshot-data`  
- If there is an existing issue this addresses, include "Fixes #XXXX" to auto-link the issue to this PR
- If there is NOT an existing issue, consider creating one.
  - In general, issues describe wanted change from an end-user perspective and PRs describe the technical change.
  - If this change is very small and not worth splitting off an issue, include `Steps To Reproduce`, `Expected Behavior`, and `Actual Behavior` sections in this PR as you would have in the issue.
- Describe what users need and how the fix will affect them
- Describe how the code change addresses the problem
- Ensure private information is redacted.
-->

## Things to be aware of

<!--
- Describe the technical choices you made
- Describe impacts on the codebase
-->

## Things to worry about

<!--
- List any questions or concerns you have with the change
- List unknowns you have 
-->

## Additional Context

<!--
Add any other context about the problem here.
-->

b-gyula committed:
XSD changes:
- remove non-root elements not allowing other root element than `databaseChangeLog`
- added documentation for some elements
- Fix: addNotNullConstraint@schemaName + @catalogName was missing
- Non-reusable attributeGroups merged into the single elements they are used
- New attributeGroups for common attributes: schemaNameAttribute
- integerExp made based on xsd:long and non-negative
- property@name restricted according to the documentation site
- required string parameters made `nonEmptyString` to avoid SQL error for definitions like
-`rowCount@expectedRows`, `setTableRemarks@remarks` made mandatory

Code changes:
- simple clear error message if the root element is other than `databaseChangeLog`: databaseChangeLog expected as root element
- 'uniqueConstraint' was misspelled

b-gyula committed:
Merge branch 'liquibase:master' into XSD.remove.invalid.root.elements

MalloD12 committed:
Merge branch 'master' into XSD.remove.invalid.root.elements

MalloD12 commented:
Hi @b-gyula,

This PR is still in draft state, is there any additional changes you would like to make here?  Updating XSD is not a trivial thing and is something that would require deeper testing. I think we are making many changes here, and it would be a bit risky moving forward with all this. I think it would be a good idea to split up this PR into smaller pieces. For one side, I think we can address the root element issue, documentation updates for one other, required parameters for another, etc. 

So far, I've seen missing attributes for some specific elements. For example, for `property` element I could see we are not including the following ones:

- relativeToChangelogFile
- errorIfMissing
- context
- contextFilter
- labels
- global

On the other hand, I could see these changes are breaking our integration tests (hence breaking some of our elements for XML changelog format).

- For example,  for `onChangeLogPreconditionErrorOrFail` type I have to add the below values:

```xml
<xsd:simpleType name="onChangeLogPreconditionErrorOrFail">
        <xsd:restriction base="xsd:string">          
            <xsd:enumeration value="MARK_RAN"/>
            <xsd:enumeration value="CONTINUE"/>
        </xsd:restriction>
    </xsd:simpleType>
```

- `Column` complex type is also broken without `mixed="true"`

-  `dropView` is missing `ifExists` attribute.

```xml
<xsd:complexType name="dropView">      
        <xsd:attribute name="ifExists" type="booleanExp"/>
</xsd:complexType>
```
- etc

b-gyula committed:
XSD changes:
- `@incrementBy`, `@minValue`, `@maxValue` changed to integer for `createSequence` and `alterSequence`
- `dropUniqueConstraint@constraintName` made required
- Invalid attributes removed from `dropSequence`
- Separate values for `@onFail` and `@onError` attributes on `preConditions` in `databaseChangeLog` and `changeSet`
- Added some documentation for entities
- Attribute group `changeAttributes` added to `catalogAndSchemaName`
- The declaration order of some attribute changed based on usefulness
- Removed duplicate `anyAttribute` definitions on some elements and added to every element were it was missing

b-gyula committed:
XSD changes:
- Add `@relativeToChangelogFile` and `@errorIfMissing` back to `property`

MalloD12 commented:
Hi @b-gyula,

As mentioned above, I think it would be good to split up this PR into smaller pieces.  We are thinking of taking part of the changes you submitted in this PR to fix #6023. Then, we can keep discussing the other refactoring you are tyring to address here. How does it sound to you?

Thanks,
Daniel.

mariochampion commented:
hey @MalloD12 lets take the smaller part of the PR we knwo will not cause issues and move that through, and if that means a whole new issue with a comment here to link them that would be fine. I ll leave it up to you how to make that happen. In the mean time i ll move this to FTT column and if we hear back from @b-gyula later we ll revive discussion of the deferred elements of the original PR. thanks!

MalloD12 commented:
@mariochampion, I took a smaller piece of this PR to solve the root element issue and added an additional piece of code that was needed. I thought this was enough, but I was breaking some existing tests that I was updating incorrectly. We discussed it with @filipelautert and agreed probably the best way to fix the root element issue would be to read the XML changelog content as a String and then verify whether the first tag element is or not a `databaseChangeLog` tag, I tried to do that on Friday without any luck yet, but I'll revisit this once I finish with the dashboard task.

Thanks,
Daniel.
